storage: use index for timequeries and support negative time deltas

VladLazar opened the pull request:
This set of patches brings the following improvements to time-queries:

1. The relative time index is used whenever possible (i.e. the batches in the segments
have monotonic timestamps). The index was previously _not used_ for time-queries.
Should be quite a bit faster, but I didn't get a chance to benchmark yet.
2. Support out of order user provided timestamps by offsetting the relative time index
by 2^31. This gives us the rough representable range of: `-596h...+596h`. This PR also
sets the default maximum time a segment can not roll for to `2 weeks`, which ensures
that the timestamps will always be in range (when users don't provide them).
3. Improve the heuristic for finding the maximum timestamp on truncation. Previously,
we simply took the last entry in the index which could be off by 64KB. Now, we take the
timestamp from the batch preceding the truncation point.

Generally, these changes make the semantics for timequery stronger:
"a timequery for timestamp 'needle' will return the first offset in the log where 'timestamp >= needle'".

Related https://github.com/redpanda-data/redpanda/issues/4035
This PR does not address the mixing of timestamps between configuration and raft batches
in the segment index. I think everything else from https://github.com/redpanda-data/redpanda/issues/4035
should be addressed.

## Backports Required

Not sure if it will be possible to backport this.

- [ ] none - not a bug fix
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [X] v22.3.x
- [ ] v22.2.x
- [ ] v22.1.x

## UX Changes

<!--

Content in this section is OPTIONAL.

Describe, in plain language, how this PR affects an end-user. Explain
topic flags, configuration flags, command line flags, deprecation
policies, etc. that are added or modified. Don't ship user breaking
changes. Ask the @redpanda-data/product team if you need help with user
visible changes.

-->

## Release Notes
### Improvements
* Handle user provided timestamps on the timequery read path more gracefully. 
* Sped up time queries by using indexing.
* Improved heuristic for determining the maximum timestamp on index truncation. This should result
in more precise time retention enforcement.

VladLazar commented:
/ci-repeat

jcsp commented:
#4035 is also tracking two other timequery gotchas:
- Segments with no user data will have walltime timestamps only, which can trip up the read path by e.g. causing it to seek very early in the log, then read a lot of data before it finds eligible user_data batches to satisfy the query.
- Truncation should take highest timestamp in data as the max timestamp, not the one that's last in the array.

Totally fine for this PR to stand alone as a fix for the intra-segment issues, but we shouldn't close out #4035 yet (this PR has a Fixes line)

VladLazar commented:
>  Segments with no user data will have walltime timestamps only, which can trip up the read path by e.g. causing it to seek very early in the log, then read a lot of data before it finds eligible user_data batches to satisfy the query.

Good point. I'll have a think about how to address this. I think this change is still a net improvement though as before we scanned the whole log in all cases.

>  Truncation should take highest timestamp in data as the max timestamp, not the one that's last in the array.

This makes sense to me. I can update this PR.

> Totally fine for this PR to stand alone as a fix for the intra-segment issues, but we shouldn't close out #4035 yet (this PR has a Fixes line)

I'll have a think about what improvements make sense to fold in here and update the cover letter accordingly.

VladLazar commented:
Changes in [force-push](https://github.com/redpanda-data/redpanda/compare/76a7bedec92a366ad6495e3cb06cba6138ee1483..d2fa4c09f69e2926fe167186c97dd37c528f0f86)
* Fix batch overshooting that was causing the ducktape tests to fail
* Added  unit tests for the issue above and others

VladLazar commented on src/v/storage/index_state.cc:
```diff
@@ -193,6 +195,10 @@ void read_nested(
     read_nested(p, st.position_index, 0U);
+
+    if (compat_version < index_state::offset_timestamps_serde_version) {
+        st.apply_offset_to_relative_time_index();
```
> When we serialize this, we are always serializing with the latest version, right?

That's right.

>  That would be fine with normal serde rules about appending fields (old redpandas would just ignore trailing data), but it's a problem here

Very good point. I'm guessing you are referring to a scenario where we'd like to go back to the previous version during an upgrade (we don't support explicit downgrades). That would be problematic. I was trying to avoid a feature flag here, but I think we need to wait for the entire cluster to be upgraded before persisting to disk with the new semantics.

VladLazar replied:
It was a pain, but it had to be done: see https://github.com/redpanda-data/redpanda/commit/e59883899287867499f2eb725955eea393d69bd0 and https://github.com/redpanda-data/redpanda/commit/e7eea26390f86213746a8fe728c73bc83bf59d62

VladLazar commented:
Changes in [force-push](https://github.com/redpanda-data/redpanda/compare/d2fa4c09f69e2926fe167186c97dd37c528f0f86..9f0b1294e98e97c4cef952c50f382bcd186b9262):
* Update `test_describe_topics_with_documentation_and_types` test with new default for `segment.ms`

VladLazar commented:
Changes in [force-push](https://github.com/redpanda-data/redpanda/compare/9f0b1294e98e97c4cef952c50f382bcd186b9262..bd9745f7e1d4825064b3771b573c2d7de06b8178):
* Fix python formatting.

dotnwat commented on src/v/storage/index_state.h:
```diff
@@ -77,9 +116,37 @@ struct index_state
+        // *In the case where lower_bound places on the first batch, we'll
+        // start the timequery from the beggining of the segment as the user
+        // data batch is always indexed.
+        return get_entry(std::max(dist - 1, decltype(dist){0}));
```
nit:

instead of having decltype and depending on relative_offset_index::difference_type being signed and implicitly relying on underflow of `dist-1` for dist=0 case (i'm guess this last part is true), you could do: `get_entry(dist > 0 : dist - 1 : 0;)` ?

VladLazar commented on src/v/storage/tests/index_state_test.cc:
```diff
@@ -7,25 +7,77 @@
 
-static storage::index_state make_random_index_state() {
+static storage::index_state
+make_random_index_state(bool with_time_ofsetting = true) {
```
Fixed.

VladLazar commented:
Changes in [force-push]:
* Only apply offsetting when the entire cluster has been upgraded. Carried the feature table on my back for a mile for this: https://github.com/redpanda-data/redpanda/pull/8509/commits/e59883899287867499f2eb725955eea393d69bd0 and  https://github.com/redpanda-data/redpanda/pull/8509/commits/e7eea26390f86213746a8fe728c73bc83bf59d62
* Fixed some bugs found in self-review
* Improved the heuristic for finding the max timestamp on index truncation: https://github.com/redpanda-data/redpanda/pull/8509/commits/2e2f6b25c48bdf90f0e4019e02fbc9ba85bd5b4e
* More tests

VladLazar commented:
Changes in [force-push](https://github.com/redpanda-data/redpanda/compare/2e2f6b25c48bdf90f0e4019e02fbc9ba85bd5b4e..f1307df54e615e9afe06cf09d6f88f5bdb5c945b):
* Noah's nit

VladLazar commented:
Changes in [force-push](https://github.com/redpanda-data/redpanda/compare/f1307df54e615e9afe06cf09d6f88f5bdb5c945b..dbbc40a13818810fd92d850b127807acb3b97c8a):
* Fixed a bug in the index state de-serialisation log that broke upgrades
* Fixed unit test failures

VladLazar commented:
Looks like all the tests passed, but stuff is hanging. Will repeat.

VladLazar commented:
/ci-repeat

VladLazar commented:
Looks like storage_single_thread_rpunit, storage_e2e_single_thread_rpunit hanged on the debug run. I'm a bit suspicious of this, so I'll run them in a loop locally and see if I can get them to hang.

VladLazar commented:
Changes in [force-push](https://github.com/redpanda-data/redpanda/compare/dbbc40a13818810fd92d850b127807acb3b97c8a..25a1787463bd11cd4f409b09924dc800b5082214):
* `stop()` stray feature tables in tests and fixtures

jcsp commented:
/ci-repeat 5

VladLazar commented:
All tests passed, but there was a botched ducktape sub-run [here](https://buildkite.com/redpanda/redpanda/builds/22594#01862746-8f40-404b-9b63-e925336de2f7).
```
Failed to run task "ci:run-redpanda-ducktape-tests-subset": task: Failed to run task "ci:build-cpp-toolchain-image": exit status 1
```

I'd say this is good to merge.

piyushredpanda commented:
@VladLazar should we not backport this?
